pkg/amiclean/ami_cleaner.go

```diff
@@ -0,0 +1,149 @@
+package amiclean
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"go.uber.org/zap"
+
+	"time"
+)
+
+const (
+	// RFC8601 is the date/time format used by AWS.
+	RFC8601 = "2006-01-02T15:04:05.000Z"
+)
+
+// AMIClean defines parameters for cleaning up AMIs based on the Branch and
+// Expiration Date.
+type AMIClean struct {
+	Delete         bool
+	Branch         string
+	Invert         bool
+	ExpirationDate time.Time
+	Logger         *zap.Logger
+	EC2Client      *ec2.EC2
+}
+
+// GetImages gets us all the private AMIs on our account so that they can be
+// looked through later. We have to do this here because the AWS API does not
+// allow you to search for AMIs by creation date or by *not* having a tag set to
+// a certain value, which would speed this up considerably.
+func (a *AMIClean) GetImages() (*ec2.DescribeImagesOutput, error) {
+	var output *ec2.DescribeImagesOutput
+
+	input := &ec2.DescribeImagesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("is-public"),
+				Values: []*string{aws.String("false")},
+			},
+		},
+	}
+
+	output, err := a.EC2Client.DescribeImages(input)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return output, err
+}
+
+// FindImagesToPurge looks through the AMIs available and produces a slice of
+// ec2.Image objects to put on the chopping block based on the contents of the
+// AMIClean struct.
+func (a *AMIClean) FindImagesToPurge(output *ec2.DescribeImagesOutput) []*ec2.Image {
+	var ImagesToPurge []*ec2.Image
+	for _, image := range output.Images {
+		imageCreationTime, _ := time.Parse(RFC8601, *image.CreationDate)
+		if imageCreationTime.After(a.ExpirationDate) {
+			continue
+		} else {
+			if a.Invert {
+				for _, tag := range image.Tags {
+					if *tag.Key == "Branch" && *tag.Value != a.Branch {
+						a.Logger.Info("selected ami for purging",
```

Comment:
I'm not sure this log line is absolutely necessary, because having all 3 log events (select, deregister and delete ami) fire creates a lot of output. I'm wondering if we could silence this one and keep the deregister and delete log lines instead? I also think it might be useful to log the number of AMIs selected, e.g. `a.Logger.Info("Selected %d AMIs to purge", len(ImagesToPurge))` right above the return line below? Up to you, but I'm just trying to imagine debugging the log lines in cloudwatch...

Reply:
I like the idea of giving a number of AMIs to purge; the log line about selecting the AMI for purging was *mostly* for my debugging and can probably go, though. I generally err on the side of *more* logging being available rather than less, but I can see how this might be redundant.

Reply:
Looking at the code for now, I'm going to leave it in, because it's the only place you can see the branch and creation time for the AMIs it wants to delete -- and getting that in the output for PurgeImages means replicating that loop. I put in a TODO to pull that functionality out so it's at least more reusable, and maybe find a more efficient way of doing it than looping though -- I'm still not sure why tags are a struct of strings instead of a map, since they seem like a perfect use case for one...